Stop adding “noreferrer” to external links.

adamsilverstein opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Fixes https://github.com/WordPress/gutenberg/issues/26914.

Removes the `noreferrer` value from the `rel` attribute on external links generated by the editor, admin UI, and block output. `noopener` is retained where needed to prevent reverse-tabnabbing on `target="_blank"` links.

### Scope

The change touches two groups of links:

- **Admin-UI links** (`ExternalLink` component, More Menus in `editor`/`edit-widgets`/`customize-widgets`, `media-categories` "view more" links, editor-only links in `site-logo` and `categories`): wp-admin sends `Referrer-Policy: strict-origin-when-cross-origin` via [`wp_admin_headers`](https://developer.wordpress.org/reference/functions/wp_admin_headers/) — which is also the modern browser default — so only the origin is sent, never full URLs or query strings. Removing `noreferrer` here does not leak sensitive admin paths.
- **Front-end post content** (`latest-posts`, `button`, `image` block output; the link format in `format-library/src/link/utils.js`; the phrasing-content paste reducer; and the associated test fixtures): sharing the referrer is the expected web-platform default, and the site's own Referrer-Policy governs what is sent.

### File block — excluded from this PR

Per [@t-hamano's review](https://github.com/WordPress/gutenberg/pull/26968#pullrequestreview-4104975430), changes to the File block have been deferred to a separate follow-up PR #77345. The File block's `save()` function deterministically generates the `rel` attribute, so changing its output invalidates existing saved File blocks on reload. The follow-up will remove both `noreferrer` AND `noopener` in one pass, with a proper new deprecation to migrate existing content cleanly.

See discussion: https://github.com/WordPress/gutenberg/issues/26914#issuecomment-4011976970

## How has this been tested?
- Unit tests and snapshots updated (`npm run test:unit`)
- E2E tests for buttons, links, and pattern overrides pass against updated `rel` output

## Types of changes
Bug fix / behavior change — removes `noreferrer` from editor-generated external links. One breaking change for `@wordpress/components` consumers: `ExternalLink` no longer adds `noreferrer` by default (documented in the components CHANGELOG).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [ ] I've included developer documentation if appropriate.
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

adamsilverstein committed:
Remove “noreferrer”.

talldan commented on packages/block-library/src/file/save.js:
```diff
@@ -20,7 +20,7 @@ export default function save( { attributes } ) {
 					<a
 						href={ textLinkHref }
 						target={ textLinkTarget }
-						rel={ textLinkTarget ? 'noreferrer noopener' : false }
```
@adamsilverstein I tested image, buttons and file blocks to see whether a deprecation is needed. It looks like the file block requires a deprecation because the save function is directly modified here.

The other two blocks will keep the existing `rel` values for existing blocks in posts and only use `noopener` for new blocks added.

adamsilverstein replied:
Ok, good catch, I haven't written a deprecation, I will give it a try though.

talldan commented on packages/e2e-tests/fixtures/blocks/core__file__new-window.html:
```diff
@@ -1,3 +1,3 @@
 <!-- wp:file {"href":"http://localhost:8888/wp-content/uploads/2018/05/keycodes.js","showDownloadButton":true} -->
-<div class="wp-block-file"><a href="http://localhost:8888/wp-content/uploads/2018/05/keycodes.js" target="_blank" rel="noreferrer noopener">6546</a><a href="http://localhost:8888/wp-content/uploads/2018/05/keycodes.js" class="wp-block-file__button" download>Download</a></div>
```
As a deprecation is needed, the practice here is to keep the old markup in a fixture called ` packages/e2e-tests/fixtures/blocks/core__file__deprecated-1.html` which will add test coverage ensuring the deprecation works. The new fixture can be added as well alongside it.

adamsilverstein replied:
Ok, will do.

adamsilverstein replied:
Hey @talldan - finally getting back to this :)

Since there is now already a `core__file__new-window__deprecated-1.html` I assume I need to create `core__file__new-window__deprecated-2.html`? do I then leave the original and first deprecation untouched?

adamsilverstein committed:
merge trunk, overwriting changes

adamsilverstein committed:
reintroduce noreferrer removal

adamsilverstein committed:
additional noreferrer removal

adamsilverstein committed:
file block test fixture noreferrer removal

adamsilverstein committed:
revert snapshot change

adamsilverstein committed:
remove fixture noreferrer references

adamsilverstein committed:
remove noreferrer from slack test fixtures

adamsilverstein committed:
adjust test fixtures

adamsilverstein committed:
update snapshots with `npm run test:unit:update`

adamsilverstein committed:
merge trunk

adamsilverstein committed:
remove some addiitonal noreferrer instances

adamsilverstein committed:
satisfy the linter for now

adamsilverstein commented:
* needs: update the linter to stop flagging corrections
* add deprecation(s) as needed, see [above](https://github.com/WordPress/gutenberg/pull/26968#discussion_r524962622)

adamsilverstein committed:
Merge trunk, resolve conflicts keeping trunk's onClick pattern

adamsilverstein committed:
Remove additional noreferrer instances from source code

Remove noreferrer from media-preview download link, block-fields
link constant (NEW_TAB_REL), and update stale snapshot.

adamsilverstein committed:
Restore noreferrer in deprecated file block save functions and fixtures

The deprecated save functions must match old saved content in the
database (which has noreferrer noopener). Only the current save
function should output noopener alone.

adamsilverstein committed:
Merge trunk

adamsilverstein committed:
Merge remote-tracking branch 'origin/trunk' into remove/noreferrer

adamsilverstein committed:
Merge remote-tracking branch 'origin/trunk' into remove/noreferrer

adamsilverstein committed:
Add CHANGELOG entry for ExternalLink noreferrer removal

adamsilverstein committed:
Fix lint errors: formatting and jsx-no-target-blank after noreferrer removal

jsnajdr reviewed (approved):
Looks good :+1: Given that we now have the global `Referrer-Policy` and that modern browsers send only origin by default, the original reasons why referrers were dangerous (leaking URL details) no longer hold, they've been solved at web platform level.

Similarly, modern browsers set `noopener` automatically for `target="blank"`, so the second ubiquitous `rel` value can be removed, too. That will also completely remove a lot of the `NEW_TAB_REL` code.

t-hamano commented on packages/block-library/src/file/save.js:
```diff
@@ -59,9 +59,7 @@ export default function save( { attributes } ) {
-						rel={
-							textLinkTarget ? 'noreferrer noopener' : undefined
-						}
+						rel={ textLinkTarget ? 'noopener' : undefined }
```
This change may break blocks, so I think we'll need to add a new migration. Try the following steps:

1. Check out the trunk branch
2. Insert a File block
3. Attach a PDF file and enable "Open in new tab"
4. Save the post 
5. Check out the `remove/noreferrer ` branch
6. Reload your browser
7. The block is broken

Perhaps we can exclude the File block in this PR. To avoid adding multiple block migrations, it might be better to submit a separate PR specifically for the File block and then remove `noreferer`, `noopener` all at once.

jsnajdr replied:
We should do a migration. Then it makes sense to remove both `noreferrer` and `noopener` in one PR.

We did a similar migration in #43050, for a similarly "banal" change--changing the value of an `aria-label` attribute. The change doesn't need to be big and serious to warrant a migration script.

Mamaduka replied:
It would be great to land #73116, so we don't have to introduce deprecations for every minor change.

adamsilverstein replied:
> Perhaps we can exclude the File block in this PR.

Good idea @t-hamano - I am going to exclude the File block for now, and open a separate PR just for it.

> We should do a migration.

I will include a migration in the new PR.

> It would be great to land #73116, so we don't have to introduce deprecations for every minor change.

@Mamaduka so after that merges, we wouldn't need a migration at all?

adamsilverstein replied:
removed changes to File block in https://github.com/WordPress/gutenberg/pull/26968/commits/2f2f030fd6df170c7d11e0b9e4b5eb9a75bf49de

adamsilverstein replied:
Follow up PR to update the File block, including migration: https://github.com/WordPress/gutenberg/pull/77345

adamsilverstein committed:
Revert File block changes; defer to follow-up PR with migration

The File block's save function deterministically generates the rel
attribute. Changing its output invalidates existing saved blocks on
reload. Defer to a follow-up PR that removes both noreferrer and
noopener together with a proper deprecation to handle migration
cleanly.

adamsilverstein committed:
Merge remote-tracking branch 'origin/trunk' into remove/noreferrer

# Conflicts:
#	packages/blocks/src/api/raw-handling/phrasing-content-reducer.ts
#	packages/components/CHANGELOG.md

adamsilverstein committed:
Merge remote-tracking branch 'origin/trunk' into remove/noreferrer

Resolve CHANGELOG conflict: preserve the ExternalLink breaking-change
note in Unreleased and keep the 32.6.0 release section from trunk.
